impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeMetricImpl.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.implcore.metrics;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import io.opencensus.common.Clock;
+import io.opencensus.common.ToDoubleFunction;
+import io.opencensus.metrics.DoubleGaugeMetric;
+import io.opencensus.metrics.LabelKey;
+import io.opencensus.metrics.LabelValue;
+import io.opencensus.metrics.export.Metric;
+import io.opencensus.metrics.export.MetricDescriptor;
+import io.opencensus.metrics.export.MetricDescriptor.Type;
+import io.opencensus.metrics.export.TimeSeries;
+import io.opencensus.metrics.export.Value;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.Nullable;
+
+/** Implementation of {@link DoubleGaugeMetric}. */
+public final class DoubleGaugeMetricImpl extends DoubleGaugeMetric implements Meter {
+  private final MetricDescriptor metricDescriptor;
+  private final ConcurrentMap<List<LabelValue>, Point> pointConcurrentHashMap =
+      new ConcurrentHashMap<List<LabelValue>, Point>();
+  private final List<LabelKey> labelKeys;
+
+  DoubleGaugeMetricImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_DOUBLE, labelKeys);
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValues, obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public Point getDefaultPoint() {
+    List<LabelValue> labelValues = new ArrayList<LabelValue>(labelKeys.size());
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  private <T> Point addAndGetPoint(
+      List<LabelValue> labelValues, @Nullable T obj, @Nullable ToDoubleFunction<T> function) {
+    Point currentPoint = pointConcurrentHashMap.get(labelValues);
```

Comment:
Here you have a race-condition and you can end with multiple PointImpl objects created. If this implementation is created this is not how should be implemented. Maybe easier just to synchronize this method for the moment and always create a new map and you keep a volatile variable for the map. See the trick here https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/metrics/export/MetricProducerManagerImpl.java#L31. With this you put more overhead into getting the point but lock free to retrieve the points.